src/Messages/OWSMessageSender.m

```diff
@@ -41,6 +41,163 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+/**
+ * OWSSendMessageOperation encapsulates all the work associated with sending a message, e.g. uploading attachments,
+ * getting proper keys,
+ * and retrying upon failure.
+ *
+ * Used by `OWSMessageSender` to serialize message sending, ensuring that messages are emitted in the order they
+ * were sent.
+ */
+@interface OWSSendMessageOperation : NSOperation
+
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithMessage:(TSOutgoingMessage *)message
+                  messageSender:(OWSMessageSender *)messageSender
+                        success:(void (^)())successHandler
+                        failure:(void (^)(NSError *_Nonnull error))failureHandler NS_DESIGNATED_INITIALIZER;
+
+@end
+
+typedef NS_ENUM(NSInteger, OWSSendMessageOperationState) {
+    OWSSendMessageOperationStateNew,
+    OWSSendMessageOperationStateExecuting,
+    OWSSendMessageOperationStateFinished
+};
+
+@interface OWSMessageSender (OWSSendMessageOperation)
+
+- (void)attemptToSendMessage:(TSOutgoingMessage *)message
+                     success:(void (^)())successHandler
+                     failure:(void (^)(NSError *error))failureHandler;
+
+@end
+
+NSString *const OWSSendMessageOperationKeyIsExecuting = @"isExecuting";
+NSString *const OWSSendMessageOperationKeyIsFinished = @"isFinished";
+
+NSUInteger const OWSSendMessageOperationMaxRetries = 4;
+
+@interface OWSSendMessageOperation ()
+
+@property (nonatomic, readonly) TSOutgoingMessage *message;
+@property (nonatomic, readonly) OWSMessageSender *messageSender;
+@property (nonatomic, readonly) void (^successHandler)();
+@property (nonatomic, readonly) void (^failureHandler)(NSError *_Nonnull error);
+@property (atomic) OWSSendMessageOperationState operationState;
+
+@end
+
+@implementation OWSSendMessageOperation
```

Comment:
It looks like this operation assume it will never be cancelled, because if it is cancelled before it is started, no handler will ever be called.  I think that's fine.